File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

vector<int> solution(vector<string> enroll, vector<string> referral, vector<string> seller, vector<int> amount) {
    vector<int> income;
    map<string, int> seq;
    
    for (int i = 0; i < enroll.size(); i++) {
        seq[enroll[i]] = i;
        income.push_back(0);
    }
    
    for (int i = 0; i < seller.size(); i++) {
        string nowSeller = seller[i];
        int nowIncome = amount[i] * 100;
        income[seq[nowSeller]] += nowIncome;
        
        while (nowSeller != "-" && nowIncome) {
            string getSeller = referral[seq[nowSeller]];
            
            nowIncome /= 10;
            if (getSeller != "-")
                income[seq[getSeller]] += nowIncome;
            income[seq[nowSeller]] -= nowIncome;
            
            nowSeller = getSeller;
        }
    }
    return income;
}

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.*;

class Solution {
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        HashMap <String, Integer> sequences = new HashMap<>();
        int n = enroll.length;
        int[] income = new int[n];
        
        // 판매책의 조직 구성원 이름 - 순서쌍 저장
        for (int i = 0; i < n; i++) 
            sequences.put(enroll[i], i);
        
        for (int i = 0; i < seller.length; i++) {
            int nowIncome = amount[i] * 100;
            String nowSeller = seller[i];
            
            while (!nowSeller.equals("-") && nowIncome > 0) {
                String nowGetter = referral[sequences.get(nowSeller)];
                
                int divide = nowIncome / 10;
                if (!"-".equals(nowGetter))
                    income[sequences.get(nowGetter)] += divide;
                
                if (!nowSeller.equals(seller[i]))
                    income[sequences.get(nowSeller)] -= divide;
                
                // System.out.println(nowSeller + " " + nowGetter + " " + divide + " " + nowIncome);
                nowIncome = divide;
                nowSeller = nowGetter;
            }
            income[sequences.get(seller[i])] += amount[i] * 90;
        }
        
        return income;
    }
}